modules/core/src/lifecycle/component.js

```diff
@@ -0,0 +1,43 @@
+import {LIFECYCLE} from '../lifecycle/constants';
+import {createProps} from '../lifecycle/create-props';
+// import {diffProps} from '../lifecycle/props';
+// import log from '../utils/log';
+// import assert from '../utils/assert';
+
+import ComponentState from './component-state';
+
+const defaultProps = {};
+
+let counter = 0;
+
+export default class Component {
+  constructor(/* ...propObjects */) {
+    // Merge supplied props with default props and freeze them.
+    /* eslint-disable prefer-spread */
+    this.props = createProps.apply(this, arguments);
+    /* eslint-enable prefer-spread */
+
+    // Define all members before layer is sealed
+    this.id = this.props.id; // The layer's id, used for matching with layers from last render cycle
+    this.count = counter++; // Keep track of how many layer instances you are generating
+    this.lifecycle = LIFECYCLE.NO_STATE; // Helps track and debug the life cycle of the layers
+    this.parentLayer = null; // reference to the composite layer parent that rendered this layer
+    this.context = null; // Will reference layer manager's context, contains state shared by layers
+    this.state = null; // Will be set to the shared layer state object during layer matching
+    this.internalState = null;
+
+    // Seal the layer
+    Object.seal(this);
+  }
+
+  // clone this layer with modified props
+  clone(newProps) {
+    return new this.constructor(Object.assign({}, this.props, newProps));
+  }
+
+  _initState() {
```

Comment:
should this be called in the constructor?

Reply:
> should this be called in the constructor?

* State initialization is part of the component lifecycle rather than construction. (`Component` will contain the lifecycle related code of `Layer`).

* The plan to move the lifecycle related parts of `LayerManager` into a `ComponentManager`.

* I didn't add that class to this PR series to keep the reviews manageable. I did include the `Component` in an attempt to make it clear where this is going.